class Solution {
    public int[] productExceptSelf(int[] nums) {
        int prod = 1;
        int zeroCount = 0;
        for (int n : nums) {
            if (n != 0) {
                prod *= n;
            }
            else {
                zeroCount++;
            }
        }

        if (zeroCount > 1) {
            return new int[nums.length];
        }

        int[] res = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            if (zeroCount > 0) {
                if (nums[i] == 0) {
                    res[i] = prod;
                }
                else {
                    res[i] = 0;
                }
            }
            else {
                res[i] = prod / nums[i];
            }
        }

        return res;
    }
}  